base/special/exp.jl

```diff
@@ -1,65 +1,248 @@
+                 0x5fed6c76e862e6d3, 0xa77d80e316c98397, 0x9a0d955d71ff6075, 0x9c2da9e603db3285, 0xa24dbe7cd63a8314,
+                 0x92ddd321f301b460, 0xa1ade7d5641c0657, 0xa72dfc97337b9b5e, 0xadae11676b197d16, 0xa42e264614f5a128,
+                 0xa30e3b333b16ee11, 0x839e502ee78b3ff6, 0xaa7e653924676d75, 0x92de7a51fbc74c83, 0xa77e8f7977cdb73f,
+                 0xa0bea4afa2a490d9, 0x948eb9f4867cca6e, 0xa1becf482d8e67f0, 0x91cee4aaa2188510, 0x9dcefa1bee615a27,
+                 0xa66f0f9c1cb64129, 0x93af252b376bba97, 0xacdf3ac948dd7273, 0x99df50765b6e4540, 0x9faf6632798844f8,
+                 0xa12f7bfdad9cbe13, 0xaeef91d802243c88, 0x874fa7c1819e90d8, 0xacdfbdba3692d513, 0x62efd3c22b8f71f1, 0x74afe9d96b2a23d9)
 
-# max and min arguments
-MAX_EXP(::Type{Float64}) = 7.09782712893383996732e2 # log 2^1023*(2-2^-52)
-MAX_EXP(::Type{Float32}) = 88.72283905206835f0      # log 2^127 *(2-2^-23)
+@inline function table_unpack(ind)
+    j = @inbounds J_TABLE[ind]
+    jU = reinterpret(Float64, JU_CONST | (j&JU_MASK))
+    jL = reinterpret(Float64, JL_CONST | (j>>8))
+    return jU, jL
+end
 
-# one less than the min exponent since we can sqeeze a bit more from the exp function
-MIN_EXP(::Type{Float64}) = -7.451332191019412076235e2 # log 2^-1075
-MIN_EXP(::Type{Float32}) = -103.97207708f0            # log 2^-150
+# Method for Float64
+# 1. Argument reduction: Reduce x to an r so that |r| <= log(b, 2)/512. Given x, base b,
+#    find r and integers k, j such that
+#       x = (k + j/256)*log(b, 2) + r,  0 <= j < 256, |r| <= log(b,2)/512.
+#
+# 2. Approximate b^r-1 by 3rd-degree minimax polynomial p_b(r) on the interval [-log(b,2)/512, log(b,2)/512].
+#    Since the bounds on r are very tight, this is sufficient to be accurate to floating point epsilon.
+#
+# 3. Scale back: b^x = 2^k * 2^(j/256) * (1 + p_b(r))
+#    Since the range of possible j is small, 2^(j/256) is stored for all possible values in slightly extended precision.
 
-@inline exp_kernel(x::Float64) = @horner(x, 1.66666666666666019037e-1,
-    -2.77777777770155933842e-3, 6.61375632143793436117e-5,
-    -1.65339022054652515390e-6, 4.13813679705723846039e-8)
+# Method for Float32
+# 1. Argument reduction: Reduce x to an r so that |r| <= log(b, 2)/2. Given x, base b,
+#    find r and integer N such that
+#       x = N*log(b, 2) + r,  |r| <= log(b,2)/2.
+#
+# 2. Approximate b^r by 7th-degree minimax polynomial p_b(r) on the interval [-log(b,2)/2, log(b,2)/2].
+# 3. Scale back: b^x = 2^N * p_b(r)
 
-@inline exp_kernel(x::Float32) = @horner(x, 1.6666625440f-1, -2.7667332906f-3)
+# For both, a little extra care needs to be taken if b^r is subnormal.
+# The solution is to do the scaling back in 2 steps as just messing with the exponent wouldn't work.
+for (func, base) in (:exp2=>Val(2), :exp=>Val(:ℯ), :exp10=>Val(10))
+    @eval begin
+        ($func)(x::Real) = ($func)(float(x))
+        @inline function ($func)(x::T) where T<:Float64
```

Comment:
Why are we now adding `@inline` to each of the `exp` functions? I'm not sure if this is appropriate, as before we did not enforce inlining of these elementary functions. Is there a good reason for this change?

Reply:
This inlining allows vectorization when combined with `@simd ivdep`. It often can yield 2x performance improvements.

Reply:
This inline before would have been useless as the old versions had way too many branches to benefit from vectorization.

Reply:
I see, makes sense

Reply:
Still this is something to be decided upon. When I inlined `sin` and `cos` it caused some compiled code to grow a lot, so I had to undo it https://github.com/JuliaLang/julia/pull/24117 (unfortunately that PR doesn't link the original issue which I believe was by @JeffBezanson ). I don't know if this applies here as well (I'm not smart enough to know if the table will also be inlined into each function that then calls `exp`).

Quick questions: when we talk about performance improvements here, are they then your new inlined versions versus the old un-inlined (?) versions? Or did you then also apply inline to the existing versions (which would only seem fair as we could also just inline the current versions).

Reply:
An array table wouldn't be duplicated, but the tuple table might. So maybe the tuple would be preferable?
The tuple only helps when using `Base.Experimental.@aliasscope`.

The array still SIMDs with `@simd ivdep`. Benchmarks on my computer, using [BetterExp.jl](https://github.com/oscardssmith/BetterExp.jl), the easiest way to test this implementation:
```julia
julia> using BetterExp, BenchmarkTools

julia> @noinline niexp(x) = BetterExp.exp(x) # noinline exp
niexp (generic function with 1 method)

julia> function simdivdepf!(f, y, x)
           @inbounds @simd ivdep for i ∈ eachindex(y,x)
               y[i] = f(x[i])
           end
       end
simdivdepf! (generic function with 1 method)

julia> x = 10randn(256); y = similar(x);

julia> @benchmark simdivdepf!(exp, $y, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.331 μs (0.00% GC)
  median time:      1.337 μs (0.00% GC)
  mean time:        1.339 μs (0.00% GC)
  maximum time:     2.431 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> @benchmark simdivdepf!(niexp, $y, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.124 μs (0.00% GC)
  median time:      1.125 μs (0.00% GC)
  mean time:        1.127 μs (0.00% GC)
  maximum time:     2.233 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> @benchmark simdivdepf!(BetterExp.exp, $y, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     192.224 ns (0.00% GC)
  median time:      192.493 ns (0.00% GC)
  mean time:        192.669 ns (0.00% GC)
  maximum time:     241.836 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     621
```
If using `10randn()`, the non-inlined version is still slightly faster. If using a narrower range of inputs (e.g., `randn()`), then the current `Base.exp` took about 1 microsecond; its performance is sensitive to the inputs.

It'd be great if `@simd` could change the inlining threshold, so that it could encourage functions defined normally (without annotations) to inline.

Reply:
I haven't tried it on this branch, but you can use the new inline-cost printer to see what might be holding this back from being automatically inlined. Demo on the current `exp`:
```julia
julia> Base.print_statement_costs(exp, (Float64,))
exp(x::T) where T<:Union{Float32, Float64} in Base.Math at special/exp.jl:74
  1 1 ── %1   = Base.bitcast(UInt64, _2)::UInt64
  1 │    %2   = Base.and_int(%1, 0x7fffffffffffffff)::UInt64
  1 │    %3   = Base.bitcast(Base.Int64, _2)::Int64
  1 │    %4   = Base.slt_int(%3, 0)::Bool
  1 │    %5   = Base.ult_int(0x40862e42fefa39ef, %2)::Bool
  0 └───        goto #12 if not %5
  1 2 ── %7   = Base.ule_int(0x7ff0000000000000, %2)::Bool
  0 └───        goto #8 if not %7
  1 3 ── %9   = Base.and_int(%2, 0x000fffffffffffff)::UInt64
  1 │    %10  = Base.sle_int(0, 0)::Bool
  1 │    %11  = Base.bitcast(UInt64, 0)::UInt64
  1 │    %12  = (%9 === %11)::Bool
  1 │    %13  = Base.and_int(%10, %12)::Bool
  0 │    %14  = Base.not_int(%13)::Bool
  0 └───        goto #5 if not %14
  0 4 ──        return NaN
  0 5 ──        goto #7 if not %4
  0 6 ──        return 0.0
  0 7 ──        return Inf
  2 8 ── %20  = Base.lt_float(709.782712893384, _2)::Bool
  0 └───        goto #10 if not %20
  0 9 ──        return Inf
  2 10 ─ %23  = Base.lt_float(_2, -745.1332191019412)::Bool
  0 └───        goto #12 if not %23
  0 11 ─        return 0.0
  2 12 ┄ %26  = Base.eq_float(_2, 1.0)::Bool
  0 └───        goto #15 if not %26
  0 13 ─        goto #15 if not true
  0 14 ─        return 2.718281828459045
  1 15 ┄ %30  = Base.ult_int(0x3fd62e42fefa39ef, %2)::Bool
  0 └───        goto #27 if not %30
  1 16 ─ %32  = Base.ult_int(%2, 0x3ff0a2b23f3bab73)::Bool
  0 └───        goto #21 if not %32
  0 17 ─        goto #19 if not %4
  1 18 ─ %35  = Base.add_float(_2, 0.6931471803691238)::Float64
  0 └───        goto #20
  1 19 ─ %37  = Base.sub_float(_2, 0.6931471803691238)::Float64
  0 20 ┄ %38  = φ (#18 => -1.9082149292705877e-10, #19 => 1.9082149292705877e-10)::Float64
  0 │    %39  = φ (#18 => %35, #19 => %37)::Float64
  0 │    %40  = φ (#18 => -1, #19 => 1)::Int64
  0 └───        goto #22
  4 21 ─ %42  = Base.mul_float(1.4426950408889634, _2)::Float64
 10 │    %43  = Base.rint_llvm(%42)::Float64
  1 │    %44  = Base.fptosi(Int64, %43)::Int64
  5 │    %45  = Base.muladd_float(%43, -0.6931471803691238, _2)::Float64
  4 └─── %46  = Base.mul_float(%43, 1.9082149292705877e-10)::Float64
  0 22 ┄ %47  = φ (#20 => %38, #21 => %46)::Float64
  0 │    %48  = φ (#20 => %39, #21 => %45)::Float64
  0 │    %49  = φ (#20 => %40, #21 => %44)::Int64
  1 │    %50  = Base.sub_float(%48, %47)::Float64
  4 │    %51  = Base.mul_float(%50, %50)::Float64
  5 │    %52  = Base.muladd_float(%51, 4.1381367970572385e-8, -1.6533902205465252e-6)::Float64
  5 │    %53  = Base.muladd_float(%51, %52, 6.613756321437934e-5)::Float64
  5 │    %54  = Base.muladd_float(%51, %53, -0.0027777777777015593)::Float64
  5 │    %55  = Base.muladd_float(%51, %54, 0.16666666666666602)::Float64
  4 │    %56  = Base.mul_float(%51, %55)::Float64
  1 │    %57  = Base.sub_float(%50, %56)::Float64
  4 │    %58  = Base.mul_float(%50, %57)::Float64
  1 │    %59  = Base.sub_float(2.0, %57)::Float64
 20 │    %60  = Base.div_float(%58, %59)::Float64
  1 │    %61  = Base.sub_float(%47, %60)::Float64
  1 │    %62  = Base.sub_float(%61, %48)::Float64
  1 │    %63  = Base.sub_float(1.0, %62)::Float64
  1 │    %64  = Base.slt_int(-52, %49)::Bool
  0 └───        goto #26 if not %64
  1 23 ─ %66  = (%49 === 1024)::Bool
  0 └───        goto #25 if not %66
 20 24 ─ %68  = $(Expr(:foreigncall, "llvm.pow.f64", Float64, svec(Float64, Float64), 0, :(:llvmcall), 2.0, 1023.0, 1023.0, 2.0))::Float64
  4 │    %69  = Base.mul_float(%63, 2.0)::Float64
  4 │    %70  = Base.mul_float(%69, %68)::Float64
  0 └───        return %70
  1 25 ─ %72  = Base.add_int(1023, %49)::Int64
  1 │    %73  = Base.bitcast(UInt64, %72)::UInt64
  1 │    %74  = Base.sle_int(0, 52)::Bool
  1 │    %75  = Base.bitcast(UInt64, 52)::UInt64
  1 │    %76  = Base.shl_int(%73, %75)::UInt64
  1 │    %77  = Base.neg_int(52)::Int64
  1 │    %78  = Base.bitcast(UInt64, %77)::UInt64
  1 │    %79  = Base.lshr_int(%73, %78)::UInt64
  1 │    %80  = Base.ifelse(%74, %76, %79)::UInt64
  1 │    %81  = Base.bitcast(Float64, %80)::Float64
  4 │    %82  = Base.mul_float(%63, %81)::Float64
  0 └───        return %82
  1 26 ─ %84  = Base.add_int(1023, 52)::Int64
  1 │    %85  = Base.add_int(%84, 1)::Int64
  1 │    %86  = Base.add_int(%85, %49)::Int64
  1 │    %87  = Base.bitcast(UInt64, %86)::UInt64
  1 │    %88  = Base.sle_int(0, 52)::Bool
  1 │    %89  = Base.bitcast(UInt64, 52)::UInt64
  1 │    %90  = Base.shl_int(%87, %89)::UInt64
  1 │    %91  = Base.neg_int(52)::Int64
  1 │    %92  = Base.bitcast(UInt64, %91)::UInt64
  1 │    %93  = Base.lshr_int(%87, %92)::UInt64
  1 │    %94  = Base.ifelse(%88, %90, %93)::UInt64
  1 │    %95  = Base.bitcast(Float64, %94)::Float64
 20 │    %96  = $(Expr(:foreigncall, "llvm.pow.f64", Float64, svec(Float64, Float64), 0, :(:llvmcall), 2.0, -53.0, -53.0, 2.0))::Float64
  4 │    %97  = Base.mul_float(%63, %95)::Float64
  4 │    %98  = Base.mul_float(%97, %96)::Float64
  0 └───        return %98
  1 27 ─ %100 = Base.ult_int(%2, 0x3e30000000000000)::Bool
  0 └───        goto #29 if not %100
  1 28 ─ %102 = Base.add_float(1.0, _2)::Float64
  0 └───        return %102
  4 29 ─ %104 = Base.mul_float(_2, _2)::Float64
  5 │    %105 = Base.muladd_float(%104, 4.1381367970572385e-8, -1.6533902205465252e-6)::Float64
  5 │    %106 = Base.muladd_float(%104, %105, 6.613756321437934e-5)::Float64
  5 │    %107 = Base.muladd_float(%104, %106, -0.0027777777777015593)::Float64
  5 │    %108 = Base.muladd_float(%104, %107, 0.16666666666666602)::Float64
  4 │    %109 = Base.mul_float(%104, %108)::Float64
  1 │    %110 = Base.sub_float(_2, %109)::Float64
  4 │    %111 = Base.mul_float(_2, %110)::Float64
  1 │    %112 = Base.sub_float(%110, 2.0)::Float64
 20 │    %113 = Base.div_float(%111, %112)::Float64
  1 │    %114 = Base.sub_float(%113, _2)::Float64
  1 │    %115 = Base.sub_float(1.0, %114)::Float64
  0 └───        return %115
```
If any of your costs don't seem appropriate then we can make adjustments.

Reply:
That's a really neat feature.

But I think the issue here is that we don't necessarily want it to inline normally, but we do if it's in an `@simd ivdep` loop, because
(a) In a loop, there are many calls, so the code size cost is amortized by a greater number of calls.
(b) The big one here: inlining lets it SIMD, so the performance difference between inlined vs not-inlined is much higher (> 5x on my computer).

Therefore, the ideal behavior would probably be something along the lines of "inline if it's inside an `@simd ivdep` loop, don't inline otherwise".

Reply:
In principle it shouldn't be too hard to adjust the threshold, because of the annotation of the loop with `Expr(:loopinfo, Symbol("julia.simdloop"), nothing)`. It might be a little sensitive to when in the optimizer pipeline that runs, though.

Reply:
I'm not sure it's pretty or preferable but you *could*  define `_exp` or `inline_exp` instead of the current `exp` and then define an `exp` (that doesn't inline) so normal users would not have this happen but advanced users could specifically ask for it. I'm above my paygrade here though :)

Reply:
Don't worry, so am I :)